File: src/report/trial_balance.rs
//! Trial balance — per-account debit/credit/net from posted entries.
//! Port of the Node `src/report/trial-balance.js`.

use crate::error::Result;
use rusqlite::{params, Connection};
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct TbAccount {
    pub code: String,
    pub name: String,
    #[serde(rename = "type")]
    pub account_type: String,
    pub debit_cents: i64,
    pub credit_cents: i64,
    pub net_cents: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct TrialBalance {
    pub accounts: Vec<TbAccount>,
    pub total_debit_cents: i64,
    pub total_credit_cents: i64,
    pub balanced: bool,
}

pub fn trial_balance(conn: &Connection, year: Option<&str>) -> Result<TrialBalance> {
    let mut stmt = conn.prepare(
        "SELECT a.code, a.name, a.type,\n\
                SUM(CASE WHEN p.amount_cents > 0 THEN p.amount_cents ELSE 0 END) AS debit_cents,\n\
                SUM(CASE WHEN p.amount_cents < 0 THEN -p.amount_cents ELSE 0 END) AS credit_cents,\n\
                SUM(p.amount_cents) AS net_cents\n\
         FROM postings p\n\
         JOIN journal_entries e ON e.id = p.entry_id\n\
         JOIN accounts a ON a.id = p.account_id\n\
         WHERE e.state = 'posted'\n\
           AND (?1 IS NULL OR substr(e.date, 1, 4) = ?1)\n\
         GROUP BY a.id\n\
         ORDER BY a.code",
    )?;
    let rows = stmt
        .query_map(params![year], |row| {
            Ok(TbAccount {
                code: row.get(0)?,
                name: row.get(1)?,
                account_type: row.get(2)?,
                debit_cents: row.get(3)?,
                credit_cents: row.get(4)?,
                net_cents: row.get(5)?,
            })
        })?
        .collect::<rusqlite::Result<Vec<_>>>()?;

    let total_debit: i64 = rows.iter().map(|r| r.debit_cents).sum();
    let total_credit: i64 = rows.iter().map(|r| r.credit_cents).sum();

    Ok(TrialBalance {
        accounts: rows,
        total_debit_cents: total_debit,
        total_credit_cents: total_credit,
        balanced: total_debit == total_credit,
    })
}

File: src/report/trial-balance.js
/**
 * bukio-cli — agent-first double-entry bookkeeping for SMEs across thirty-one jurisdictions.
 * Copyright (c) 2026 Erik van Kempen.
 * SPDX-License-Identifier: Apache-2.0
 */

// Trial balance — per-account debit/credit/net from posted entries.
import { fiscalYearWindow } from '../year-end/index.js';

export function trialBalance(db, { year = null } = {}) {
  if (year != null && !/^\d{4}$/.test(String(year))) {
    const e = new Error(`year '${year}' must be YYYY`);
    e.code = 'INVALID_YEAR';
    throw e;
  }
  const [from, to] = year != null ? fiscalYearWindow(db, String(year)) : [null, null];
  const rows = db.prepare(`
    SELECT a.code, a.name, a.type,
      SUM(CASE WHEN p.amount_cents > 0 THEN p.amount_cents ELSE 0 END) AS debit_cents,
      SUM(CASE WHEN p.amount_cents < 0 THEN -p.amount_cents ELSE 0 END) AS credit_cents,
      SUM(p.amount_cents) AS net_cents
    FROM postings p
    JOIN journal_entries e ON e.id = p.entry_id
    JOIN accounts a ON a.id = p.account_id
    WHERE e.state = 'posted'
      AND (? IS NULL OR e.date >= ?)
      AND (? IS NULL OR e.date <= ?)
    GROUP BY a.id
    ORDER BY a.code
  `).all(from, from, to, to);

  const totalDebit = rows.reduce((s, r) => s + r.debit_cents, 0);
  const totalCredit = rows.reduce((s, r) => s + r.credit_cents, 0);

  return {
    accounts: rows,
    total_debit_cents: totalDebit,
    total_credit_cents: totalCredit,
    balanced: totalDebit === totalCredit,
  };
}
